relays/bin-substrate/src/chains/statemine.rs

```diff
@@ -0,0 +1,34 @@
+// Copyright 2019-2021 Parity Technologies (UK) Ltd.
+// This file is part of Parity Bridges Common.
+
+// Parity Bridges Common is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Parity Bridges Common is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Parity Bridges Common.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Statemine chain specification for CLI.
+
+use crate::cli::CliChain;
+use relay_statemine_client::Statemine;
+use sp_version::RuntimeVersion;
+
+impl CliChain for Statemine {
+	const RUNTIME_VERSION: Option<RuntimeVersion> = Some(bp_statemine::VERSION);
```

Comment:
@svyatonik 
I guess, we dont really need `bp_statemine::VERSION`, right?
for BridgeHubs we set this to `None` as `const RUNTIME_VERSION: Option<RuntimeVersion> = None;`

Reply:
No, we need it because we're going to submit transactions to Statemine. Each transaction has so called "signed extension" and to craft it you need to know actual chain version. We have three ways to sepcify this version: one is to use version that is bundled with the `substrate-relay` binary (you're looking at it now). Other option is to specify it using CLI. And the last one is to read it from the node.

The latter is ok for test chains, but for real chains we don't want to use it because call encoding can change and since we're using mocked runtime our transaction may just be rejected (and we'll probably waste our funds on that). The first option is for the production - we're going to check that our mocked runtime is not broken on every Statemine upgrade and redeploy our relay with updated version (at least that was the initial plan - feel free to suggest improvements). The 2nd option (specify version using CLI) is a temporary measure - e.g. to run relay if we sure that our call encoding is matching upgraded runtime version, but we still have not built updated relay.

Reply:
so lets say for BridgeHubRococo/Wococo we are ok with None (read from node), but for Kusama/Polkadot we will need to also setup VERSION: RuntimeVersion?

Reply:
I am just thinking loud and what to understand in deep:
- lets say we are running real Kusama/Polkadot bridges and there is Statemint/e in some versions (STv123 and SEv456),
- afaik when `transaction_version` is changed, then we need to probably check/update encoding for our mocked runtime calls
- Statemint can be upgraded just throught proposal/governance (same for bridge-hubs) and it takes "some time" (iirc cca 2 weeks) and after this period runtime upgrade is approved and new runtime is changed in running collators (lets call it TIME_OF_REAL_UPGRADE_IN_NETWORK)

---

as you said:
> we're going to check that our mocked runtime is not broken on every **Statemine upgrade** and **redeploy**

1. we can check **Statemine upgrade** - there is a Cumulus release process of runtimes, and we can check manually (e.g. if transaction_version is changed or not...), there is no problem, I guess
2. after release process, there is a proposal/governance phase for upgrading runtime
3. after governance is done, at some time (TIME_OF_REAL_UPGRADE_IN_NETWORK) **Statemine redeploy** occurrs (afaik it is triggered by user action, any user can do enact of approved runtime upgrade)
4. substrate-relay have two possibilities:
4.1. **no need to update** mocked runtime (e.g. transaction_version was not change)
4.2. **need to update** mocked runtime, at some point we need to redeploy new version (TIME_OF_REAL_UPGRADE_OF_SUBSTRATE_RELAYER) (lets say, redeploy new version, means just to push new tag)

this all also will apply for BridgeHub upgrades.

---
**Questions/open points:**
1. does initial plan count with some monitoring or automatization?
- do we need anything, e.g. for checking that Statemin/BridgeHub runtimes are redeployed?

2. **what about potential transactions, which could be rejected between TIME_OF_REAL_UPGRADE_IN_NETWORK and TIME_OF_REAL_UPGRADE_OF_SUBSTRATE_RELAYER?**
- even if we push new tag immediatelly after TIME_OF_REAL_UPGRADE_IN_NETWORK, there is still some gap till new relayer is redeployed and again start to submit extrinsics
- iiuc messages from **outbouned_queue** are removed on `receive_messages_delivery_proof`
  - could there be any problems, that messages stay too long (few hours/days) in **outbouned_queue**?
    - OutboundLane is fulfilled? I guess, now it is unbouned? also for InboundQueue?
  - what if we restart relayer, which continues to sync headers (there is a `HeadersToKeep`), could we miss/fail to execute some outbound messages on the other side because of too low `HeadersToKeep`?

are there any known issues with redeploy?